### ISSUE TITLE
Proposals: Multi-Entries Support

### ISSUE THREAD

**Issue by antfu** [slidevjs/slidev#53]
Currently, Slidev only supports a single Markdown entry for all your slides. I believe this way is simply and easier to focus on the slides content for getting started. But when the slides goes bigger can more complex, sometime you might found it hard to find the right slides to edit. We do have [editor supports](https://sli.dev/guide/editors.html) could solve this problem to some extent.

But eventually, we might still need to have multi-entries support for better organizations and reusability in some scenarios. So, this issue brings two proposals for the interface of this feature.

# Proposal A

Using frontmatter option `src` to specify the sub-entry:

```markdown
---
src: ./welcome.md
---

<!-- when `src` is used, the inline content will be ignored -->

---
src: ./intro.md
---

---

# Inline Slide

---
src: ./intro.md
layout: custom
---

<!-- resuse same slide content, but with different layout -->
```

The if the subentry contains frontmatter as well, it will be merged with the inline ones (inline frontmatter has high priority).

We could also make our editor support read / write to the correct sub-entries with this approach.

# Proposal B

"Magic Component" - use sub-entires as components, so you can also extend the content (will be replaced compile time)

```markdown
<SlideFragment src="./welcome.md" />

---

<SlideFragment src="./intro.md" />

---

# Inline Title

Inline Content

<!-- blend with inline content -->
<SlideFragment src="./page-3.md" />

---
```

This is more flexible but it's also saying we are not likely to have good editor support for sub-entries.

# Proposal C

Provide by @bramceulemans 

Having each slide as a file, under './slides' folder and follow some naming conversion. For example

```
├── slides
|   ├── 01.md
|   ├── 02.md
    └── ...
```

Ideally, we could also provide a CLI component to split into "multiple files style" and re-compose them back to a single entry when need.

The downside of this approach is losing the ability to "reuse" sections and make the re-ordering harder without additional tools.

**Comment by BellemareMederic** [comment]
Does a folder named /slides that implemented auto ordering like Hugo CMS can be more comprehensive ?
https://gohugo.io/templates/lists/#example-project-directory
like 
slide-01.md
slide-02.md
etc

**Comment by antfu** [comment]
Yeah, could do that. But I am a bit concern about how you reorder them, let's say if you want to insert move 13.md to the front, which means you will need to rename all the 13 files before it.

**Comment by meteorlxy** [comment]
> Yeah, could do that. But I am a bit concern about how you reorder them, let's say if you want to insert move 13.md to the front, which means you will need to rename all the 13 files before it.

What about a special file for ordering? E.g. `slides.yml`:

```
├── slides.yml
├── slides
|   ├── 01-foo.md
|   ├── 02-bar.md
|   ├── 03-baz.md
    └── ...
```

```yml
# slides.yml
# change slides order in this file
slides:
  - 02-bar.md
  - 03-baz.md
  - 01-foo.md
```

The drawback is that the filename order could be inconsistent with the actual order.

**Comment by meteorlxy** [comment]
> Ideally, we could also provide a CLI component to split into "multiple files style" and re-compose them back to a single entry when need.
> The downside of this approach is losing the ability to "reuse" sections and make the re-ordering harder without additional tools.

Or provide an extra cli command to help reordering slides?

**Comment by Marr11317** [comment]
I actually don't think separating into multiple files would be *that* useful. Let's say a big presentation has 1 slides (and that's starting to be a quite big presentation). 

That would make what? a thousand LOC's? That's still not that bad.

The solution might be in the editor support. What about grouping sections, to navigate quickly? Or even virtual files: you were talking about separating in files? What if opening one of these virtual files simply opened the editor at the right slides, with every other slides hidden?

If there's dire need for importing other files, both proposal A and B would be fine, but I'd hate C, since it would be the less versatile (what about extending a slide) would create the need for file navigation in small projects, when it's not needed, etc.

A would be more user friendly, B more powerful.

**Comment by draptik** [comment]
I would really love separating into multiple files! (I currently use reveal-md just for that feature)
When giving workshops lasting more than a day it is normal to have >100 slides.
Having a "main" table-of-content (TOC) file is very convenient. It allows to insert repeating slides ("waiting", "lunchbreack", "exercise").
I wouldn't need recursive nesting: 1 layer would be great though!

**Comment by filipebraida** [comment]
I really liked the slidev. I used pandoc and a script to convert my class slides to pdf. I usually created one project per subject and had several files per lesson content. I also see utility in functionality, because in this case, I would need to create a project by theme or put it together in a large confusing file. 

In this case, I like "Proposal A".

A feature that would be interesting is to give the file name for conversion.

**Comment by Bktero** [comment]
Hi! I have just tried out slidev today, very promosing :)

Being able to have several `*.md` would be a great improvement!

Solution A is my favorite. It's very simple and easy to understand.

Solution B seems very similar but it is not as clear as solution A for people who are not Web developpers (I am one of them!).

Solution C has drawbacks:

- it is not easy to reorder the slides
- it is not possible to include the same slide twice in the presentation

**Comment by antfu** [comment]
Released v0.15.0 with the **Proposal A**'s multi-entries support 🎉

- Integrated editor should save changes to the correct file, so as the "open in VS Code" button.
- HMR should also works.
- VS Code extension support will catch up later.
- Docs in a few hours.

For Proposal B, you can use `.md` as components (partial support)

```
/components
  /foo.md
```

```md
# Page

<Foo/>
```

For Proposal C, we could add some command to generate `slides.md` from file conventions in the Proposal A style in the future.

This feature is pretty much experimental, if you encountered any bugs, please open new issues. (I don't read closed issues)

**Comment by Bktero** [comment]
I can't make it work properly. Should I create new issues and explain my problems here?
